### ISSUE TITLE
Error when calling job_collect for hyperparameter tuning job

### ISSUE THREAD

**Issue by jjallaire** [rstudio/cloudml#102]
I am seeing this error when calling `job_collect()`:

```
> job_collect()
Error in order(sapply(status$trainingOutput$trials, function(e) e$finalMetric$objectiveValue),  : 
  unimplemented type 'list' in 'orderVector1'
In addition: Warning message:
In max(job_list_trials(job)) :
  no non-missing arguments to max; returning -Inf
```

I am using the script here: https://github.com/rstudio/cloudml/tree/master/inst/examples/keras

My training job is submitted with:

```r
cloudml_train("mnist_mlp.R", cloudml = "tuning.yml")
```

**Comment by javierluraschi** [comment]
Just noticed this one, will send a PR shortly...

**Comment by javierluraschi** [comment]
The TensorFlow run has status:

```
Browse[2]> status
 $ createTime    : chr "2018-01-04T21:51:22Z"
 $ endTime       : chr "2018-01-04T21:59:08Z"
 $ etag          : chr "YFZ_TgitnRs="
 $ jobId         : chr "cloudml_2018_01_04_215108244"
 $ startTime     : chr "2018-01-04T21:51:26Z"
 $ state         : chr "SUCCEEDED"
 $ trainingInput :List of 4
  ..$ hyperparameters:List of 5
  .. ..$ goal                   : chr "MAXIMIZE"
  .. ..$ hyperparameterMetricTag: chr "accuracy"
  .. ..$ maxParallelTrials      : int 2
  .. ..$ maxTrials              : int 2
  .. ..$ params                 :List of 1
  .. .. ..$ :List of 5
  .. .. .. ..$ maxValue     : num 0.5
  .. .. .. ..$ minValue     : num 0.4
  .. .. .. ..$ parameterName: chr "gradient-descent-optimizer"
  .. .. .. ..$ scaleType    : chr "UNIT_LINEAR_SCALE"
  .. .. .. ..$ type         : chr "DOUBLE"
  ..$ jobDir         : chr "gs://rstudio-cloudml/r-cloudml/staging"
  ..$ region         : chr "us-central1"
  ..$ runtimeVersion : chr "1.4"
 $ trainingOutput:List of 4
  ..$ completedTrialCount      : chr "2"
  ..$ consumedMLUnits          : num 0.19
  ..$ isHyperparameterTuningJob: logi TRUE
  ..$ trials                   :List of 2
  .. ..$ :List of 3
  .. .. ..$ finalMetric    :List of 2
  .. .. .. ..$ objectiveValue: num 0.91
  .. .. .. ..$ trainingStep  : chr "1000"
  .. .. ..$ hyperparameters:List of 1
  .. .. .. ..$ gradient-descent-optimizer: chr "0.40229339483728294"
  .. .. ..$ trialId        : chr "2"
  .. ..$ :List of 3
  .. .. ..$ finalMetric    :List of 2
  .. .. .. ..$ objectiveValue: num 0.84
  .. .. .. ..$ trainingStep  : chr "999"
  .. .. ..$ hyperparameters:List of 1
  .. .. .. ..$ gradient-descent-optimizer: chr "0.4890691043624778"
  .. .. ..$ trialId        : chr "1"

Hyperparameter Trials:
  finalMetric.objectiveValue finalMetric.trainingStep gradient.descent.optimizer trialId
1                       0.91                     1000        0.40229339483728294       2
2                       0.84                      999         0.4890691043624778       1

View job in the Cloud Console at:
https://console.cloud.google.com/ml/jobs/cloudml_2018_01_04_215108244?project=rstudio-cloudml

View logs at:
https://console.cloud.google.com/logs?resource=ml.googleapis.com%2Fjob_id%2Fcloudml_2018_01_04_215108244&project=rstudio-cloudml
```

However, the keras run:

```
Browse[2]> status
 $ createTime    : chr "2018-01-05T21:26:51Z"
 $ endTime       : chr "2018-01-05T21:43:25Z"
 $ etag          : chr "GnVF47a3VLA="
 $ jobId         : chr "cloudml_2018_01_05_212634662"
 $ startTime     : chr "2018-01-05T21:26:55Z"
 $ state         : chr "SUCCEEDED"
 $ trainingInput :List of 4
  ..$ hyperparameters:List of 5
  .. ..$ goal                   : chr "MAXIMIZE"
  .. ..$ hyperparameterMetricTag: chr "accuracy"
  .. ..$ maxParallelTrials      : int 2
  .. ..$ maxTrials              : int 2
  .. ..$ params                 :List of 1
  .. .. ..$ :List of 5
  .. .. .. ..$ maxValue     : num 0.5
  .. .. .. ..$ minValue     : num 0.4
  .. .. .. ..$ parameterName: chr "dropout-rate"
  .. .. .. ..$ scaleType    : chr "UNIT_LINEAR_SCALE"
  .. .. .. ..$ type         : chr "DOUBLE"
  ..$ jobDir         : chr "gs://rstudio-cloudml/r-cloudml/staging"
  ..$ region         : chr "us-central1"
  ..$ runtimeVersion : chr "1.4"
 $ trainingOutput:List of 4
  ..$ completedTrialCount      : chr "2"
  ..$ consumedMLUnits          : num 0.19
  ..$ isHyperparameterTuningJob: logi TRUE
  ..$ trials                   :List of 2
  .. ..$ :List of 2
  .. .. ..$ hyperparameters:List of 1
  .. .. .. ..$ dropout-rate: chr "0.41349334911118218"
  .. .. ..$ trialId        : chr "1"
  .. ..$ :List of 2
  .. .. ..$ hyperparameters:List of 1
  .. .. .. ..$ dropout-rate: chr "0.487533279123358"
  .. .. ..$ trialId        : chr "2"

Hyperparameter Trials:
         dropout.rate trialId
1 0.41349334911118218       1
2   0.487533279123358       2

View job in the Cloud Console at:
https://console.cloud.google.com/ml/jobs/cloudml_2018_01_05_212634662?project=rstudio-cloudml

View logs at:
https://console.cloud.google.com/logs?resource=ml.googleapis.com%2Fjob_id%2Fcloudml_2018_01_05_212634662&project=rstudio-cloudml
```

From the docs [HyperparameterOutput](https://cloud.google.com/ml-engine/reference/rest/v1/projects.jobs#HyperparameterOutput) should contain a `finalMetric` that we use to order and get the best one, but it's not there.

So in this case, one could argue that there is no info Keras is providing to choose the best run, is there a way we can make Keras provide this event? In the TF script, we add this event manually as `tf$summary$scalar("accuracy", accuracy)`.

In the meantime, I'll send a PR to improve the error message in this case.

**Comment by javierluraschi** [comment]
Sent https://github.com/rstudio/cloudml/pull/103 to improve error messages.
